torch/_dynamo/guards.py

```diff
@@ -496,13 +497,42 @@ def NN_MODULE(self, guard: Guard):
         else:
             self._produce_guard_code(guard, [f"{name}({ref})"])
 
+    def CODE_MATCH(self, guard: Guard):
+        """Checks that the `__code__` of two functions are equal"""
+        ref = self.arg_ref(guard)
+        # Cannot guard with local var deps
+        if re.match(r"^(.*)L\[.*\](.*)$", ref) or "L[" in ref:
+            return
+        value = self.get(guard.name)
+
+        if hasattr(value, "__code__") and isinstance(value.__code__, types.CodeType):
+            code = list()
+            code.append(f"hash({ref}.__code__.co_code) == {hash(value.__code__.co_code)}")
+            code.append(f"{ref}.__code__.co_name == '{value.__code__.co_name}'")
+            code.append(f"___basename({ref}.__code__.co_filename)"
+                        f" == '{os.path.basename(value.__code__.co_filename)}'")
+
+            # TODO(jon-chuang): Below may be too strict for export?
+            # If we want to ensure recompilation if the users' library version
+            # changes, then we'd uncomment below. Else, lineno is probably too strict.
+            # filename basename might still be fine, but full path is probably too strict.
```

Comment:
the reasoning behind this is:
It's likely for function to be same across library versions even if their line number changes due to independent changes to file.

Further, the library install path may change, but unlikely that the file itself changes.

In most cases, just checking `hash(co_code)` is sufficient, these are extra checks.